src/workflow/StatedWorkflow.js

```diff
@@ -467,16 +464,35 @@ export class StatedWorkflow {
 
         const {type:clientType} = clientParams;
 
+        // testData may contain some canned data to be used without a publisher.
         if (testData !== undefined) {
             this.logger.debug(`No 'real' subscription created because testData provided for subscription params ${StatedREPL.stringify(subscriptionParams)}`);
-            const dispatcher = this.workflowDispatcher.getDispatcher(subscriptionParams);
+            const testDataAckFunctionGenerator = ((data) => {
+                return async () => {
+                    if (subscriptionParams !== undefined && Array.isArray(subscriptionParams.acks)) {
+                        await this.templateProcessor.setData(subscribeParamsJsonPointer + '/acks/-',data);
+                    }
+                }
+            }).bind(this);
+            const dispatcher = this.workflowDispatcher.getDispatcher(subscriptionParams, testDataAckFunctionGenerator);
+
             await dispatcher.addBatch(testData);
             await dispatcher.drainBatch(); // in test mode we wanna actually wait for all the test events to process
             return;
         }
+        // clientType test means that the data will be sent directly from publish function to the dispatcher
         if(clientType==='test'){
             this.logger.debug(`No 'real' subscription created because client.type='test' set for subscription params ${StatedREPL.stringify(subscriptionParams)}`);
-            const dispatcher = this.workflowDispatcher.getDispatcher(subscriptionParams); // we need a dispatched even if no real message bus
+            const testDataAckFunctionGenerator = ((data) => {
+                return async () => {
+                    if (subscriptionParams !== undefined && Array.isArray(subscriptionParams.acks)) {
+                        console.debug(`acknowledging data: ${StatedREPL.stringify(data)}`);
+                        await this.templateProcessor.setData(subscribeParamsJsonPointer + '/acks/-',data);
+                    }
+                }
+            }).bind(this);
```

Comment:
arrow functions should never need to be bound

Reply:
removed